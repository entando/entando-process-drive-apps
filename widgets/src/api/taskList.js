import { IS_MOCKED_API, MOCK_API_DELAY, SERVICE } from 'api/constants';
import getTasks from 'mocks/taskList/api.mock';
import utils from 'utils';

const get = async (connection, page, pageSize, sortedColumn, sortOrder, filter) => {
  if (IS_MOCKED_API) {
    await utils.timeout(MOCK_API_DELAY);
    return getTasks(page, pageSize, sortedColumn, sortOrder, filter);
  }

  const url = `${SERVICE.URL}/connections/${connection}/tasks?sort=taskId`;
  const token = localStorage.getItem('token');
  const response = await fetch(url, {
    method: 'get',
    headers: new Headers({
      Authorization: `Bearer ${token}`,
      'Content-Type': 'application/json',
    }),
  });

  return response.json();
};

export default {
  get,
};
