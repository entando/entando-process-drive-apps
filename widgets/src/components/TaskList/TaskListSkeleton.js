import React from 'react';
import PropTypes from 'prop-types';
import Skeleton from '@material-ui/lab/Skeleton';
import withStyles from '@material-ui/core/styles/withStyles';

const styles = {
  header: {
    display: 'flex',
    justifyContent: 'space-between',
    padding: '10px 20px',
    '&>div': {
      width: '50%',
    },
  },
  search: {
    width: '20% !important',
    marginTop: 20,
  },
  tableRows: {
    margin: '40px auto 40px auto',
  },
};

const getKey = (qty, i) => qty - i;

const TaskListSkeleton = ({ classes, rows }) => (
  <>
    <div className={classes.header}>
      <div>
        <Skeleton width="40%" />
        <Skeleton width="20%" />
      </div>
      <div className={classes.search}>
        <Skeleton width="100%" height={40} variant="rect" />
      </div>
    </div>
    <div>
      <Skeleton width="100%" height={40} variant="rect" />
      {[...Array(rows)].map((row, i) => (
        <Skeleton key={getKey(rows, i)} className={classes.tableRows} width="98%" height={10} />
      ))}
    </div>
    <div>
      <Skeleton width="100%" height={50} variant="rect" />
    </div>
  </>
);

TaskListSkeleton.propTypes = {
  classes: PropTypes.shape({
    header: PropTypes.string,
    search: PropTypes.string,
    tableRows: PropTypes.string,
  }),
  rows: PropTypes.number,
};

TaskListSkeleton.defaultProps = {
  classes: {},
  rows: 10,
};

export default withStyles(styles)(TaskListSkeleton);
