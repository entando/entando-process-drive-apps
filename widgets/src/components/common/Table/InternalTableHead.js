import PropTypes from 'prop-types';
import React from 'react';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';

import columnType from 'types/columnType';
import InternalTableHeader from 'components/common/Table/InternalTableHeader';

export default function InternalTableHead({ columns, createSortHandler, sortedColumn, sortOrder }) {
  return (
    <TableHead>
      <TableRow>
        {columns.map(column => (
          <InternalTableHeader
            key={JSON.stringify(column)}
            column={column}
            createSortHandler={createSortHandler}
            sortedColumn={sortedColumn}
            sortOrder={sortOrder}
          />
        ))}
      </TableRow>
    </TableHead>
  );
}

InternalTableHead.propTypes = {
  columns: PropTypes.arrayOf(columnType).isRequired,
  createSortHandler: PropTypes.func,
  sortedColumn: PropTypes.string,
  sortOrder: PropTypes.string,
};

InternalTableHead.defaultProps = {
  createSortHandler: () => () => {},
  sortedColumn: '',
  sortOrder: 'asc',
};
