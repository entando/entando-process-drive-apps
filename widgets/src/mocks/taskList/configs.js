export const CONNECTIONS = {
  payload: [
    {
      name: 'kieStaging',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
    {
      name: 'kie2Server',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
    {
      name: 'test',
      host: 'rhpam7-install-kieserver-rhpam7-install-entando.apps.serv.run',
      port: '80',
      schema: 'http',
      app: '/services/rest/',
      username: 'pamAdmin',
      connectionTimeout: 60000,
      engine: 'pam',
    },
  ],
  errors: {},
};

export const PROCESS = {
  payload: [
    {
      reusableSubProcesses: null,
      'process-id': 'Mortgage_Process.MortgageApprovalProcess',
      associatedEntities: null,
      'container-id': 'mortgage-process_1.0.0-SNAPSHOT',
      package: 'com.myspace.mortgage_app',
      'process-name': 'MortgageApprovalProcess',
      dynamic: false,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1.0',
    },
    {
      reusableSubProcesses: null,
      'process-id': 'evaluation',
      associatedEntities: null,
      'container-id': 'evaluation_1.0.0-SNAPSHOT',
      package: 'Evaluation.src.main.resources',
      'process-name': 'Evaluation',
      dynamic: false,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1',
    },
    {
      reusableSubProcesses: null,
      'process-id': 'itorders.orderhardware',
      associatedEntities: null,
      'container-id': 'itorders_1.0.0-SNAPSHOT',
      package: 'org.jbpm.demo.itorders',
      'process-name': 'Order for IT hardware',
      dynamic: true,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1.0',
    },
    {
      reusableSubProcesses: null,
      'process-id': 'Mortgage_Process.MortgageApprovalProcess',
      associatedEntities: null,
      'container-id': 'mortgage-process_1.0.1-SNAPSHOT',
      package: 'com.myspace.mortgage_app',
      'process-name': 'MortgageApprovalProcess',
      dynamic: false,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1.0',
    },
    {
      reusableSubProcesses: null,
      'process-id': 'com.redhat.bpms.examples.mortgage.MortgageApplication',
      associatedEntities: null,
      'container-id': 'mortgage_1.0',
      package: 'com.redhat.bpms.examples.mortgage',
      'process-name': 'MortgageApplication',
      dynamic: false,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1.0',
    },
    {
      reusableSubProcesses: null,
      'process-id': 'itorders-data.place-order',
      associatedEntities: null,
      'container-id': 'itorders_1.0.0-SNAPSHOT',
      package: 'org.jbpm.demo.itorders',
      'process-name': 'place-order',
      dynamic: false,
      serviceTasks: null,
      processVariables: null,
      'process-version': '1.0',
    },
  ],
  errors: [],
};

export const GROUPS = {
  payload: [
    'Administrators',
    'broker',
    'admin',
    'rest-all',
    'kiemgmt',
    'kie-server',
    'appraiser',
    'manager',
    'supplier',
    'IT',
    'PM',
    'approver',
  ],
  errors: {},
};

export const COLUMNS = {
  payload: [
    'id',
    'name',
    'description',
    'createdBy',
    'createdAt',
    'dueTo',
    'status',
    'owner',
    'priority',
    'subject',
    'activatedAt',
    'skipable',
    'processDefinitionId',
    'processId',
    'parentId',
  ],
  errors: {},
};

export const WIDGET_CONFIGS = {
  payload: {
    code: 'pda_task_list',
    config: {
      groups:
        '[{"label":"Administrators","key":"Administrators","checked":true},{"label":"broker","key":"broker","checked":true},{"label":"admin","key":"admin","checked":false},{"label":"rest-all","key":"rest-all","checked":true},{"label":"kiemgmt","key":"kiemgmt","checked":false},{"label":"kie-server","key":"kie-server","checked":true},{"label":"appraiser","key":"appraiser","checked":false},{"label":"manager","key":"manager","checked":true},{"label":"supplier","key":"supplier","checked":true},{"label":"IT","key":"IT","checked":true},{"label":"PM","key":"PM","checked":true},{"label":"approver","key":"approver","checked":true}]',
      options:
        '[{"key":"newPageOnClick","label":"Open new page on table row click","checked":false},{"key":"showClaim","label":"Show Claim Button","checked":true},{"key":"showComplete","label":"Show Complete Button","checked":true}]',
      columns:
        '[{"name":"id","position":0,"isVisible":true},{"name":"name","position":1,"isVisible":true},{"name":"createdBy","position":2,"isVisible":true},{"name":"createdAt","position":3,"isVisible":true},{"name":"dueTo","position":4,"isVisible":true},{"name":"status","position":5,"isVisible":true},{"name":"description","position":6,"isVisible":true},{"name":"owner","position":7,"isVisible":true},{"name":"inputData","position":8,"isVisible":true},{"name":"outputData","position":9,"isVisible":true},{"name":"priority","position":10,"isVisible":true},{"name":"subject","position":11,"isVisible":true},{"name":"type","position":12,"isVisible":true},{"name":"form","position":13,"isVisible":true},{"name":"activatedAt","position":14,"isVisible":true},{"name":"skipable","position":15,"isVisible":true},{"name":"workItemId","position":16,"isVisible":true},{"name":"processId","position":17,"isVisible":true},{"name":"processDefinitionId","position":18,"isVisible":true},{"name":"parentId","position":19,"isVisible":true},{"name":"slaCompliance","position":20,"isVisible":true},{"name":"slaDueTo","position":21,"isVisible":true},{"name":"potentialOwners","position":22,"isVisible":true},{"name":"excludedOwners","position":23,"isVisible":true},{"name":"businessAdmins","position":24,"isVisible":true}]',
      knowledgeSource: 'kieStaging',
      process: 'Mortgage_Process.MortgageApprovalProcess@mortgage-process_1.0.1-SNAPSHOT',
    },
  },
  errors: {},
};
