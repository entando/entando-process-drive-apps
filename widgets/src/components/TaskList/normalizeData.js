import ActionCell from 'components/common/Table/custom/ActionCell';
import { compareDates, compareNumbers, compareStrings } from 'components/common/Table/utils';

export const getType = (column, firstRow) => {
  let sortFunction = compareStrings;
  if (firstRow[column] instanceof Date) {
    sortFunction = compareDates;
  } else if (typeof firstRow[column] === 'number') {
    sortFunction = compareNumbers;
  }

  return sortFunction;
};

export const normalizeColumns = (columns, firstRow) => {
  const normalized = columns
    .filter(column => column.isVisible)
    .map(column => ({
      header: column.header,
      accessor: column.name,
      position: column.position,
      sortFunction: getType(column, firstRow),
    }));
  // order columns
  normalized.sort((a, b) => (a.position > b.position ? 1 : a.position < b.position ? -1 : 0));

  // add action field
  normalized.unshift({
    header: 'Action',
    accessor: 'action',
    customCell: ActionCell,
    styles: {
      position: 'sticky',
      left: 0,
      background: 'white',
      width: 20,
      zIndex: 100,
      borderRight: '1px solid #eee',
      paddingRight: 16,
      textAlign: 'center',
    },
  });

  return normalized;
};

export const normalizeRows = rows =>
  rows.map(row => {
    const normalizedRow = {};
    Object.keys(row).forEach(key => {
      if (row[key] instanceof Object) {
        normalizedRow[key] = '';
      } else {
        normalizedRow[key] = String(row[key]);
      }
    });
    return normalizedRow;
  });
