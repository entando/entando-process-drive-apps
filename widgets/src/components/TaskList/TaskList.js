import { ThemeProvider } from '@material-ui/styles';
import React from 'react';
import i18next from 'i18next';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import withStyles from '@material-ui/styles/withStyles';

import { WIDGET_CODES } from 'api/constants';
import configApi from 'api/config';
import taskListApi from 'api/taskList';
import utils from 'utils';

import { normalizeColumns, normalizeRows } from 'components/TaskList/normalizeData';
import ErrorNotification from 'components/common/ErrorNotification';
import ErrorComponent from 'components/common/ErrorComponent';
import Table from 'components/common/Table/Table';
import theme from 'theme';

const styles = {
  paper: {
    minHeight: 459,
    position: 'relative',
  },
};

class TaskList extends React.Component {
  state = {
    loading: true,
    columns: [],
    rows: [],
    size: 0,
    currentPage: 0,
    connection: {},
    error: null,
    errorAlert: null,
  };

  timer = { ref: null };

  componentDidMount = async () => {
    const { lazyLoading } = this.props;
    const { currentPage } = this.state;

    try {
      const config = await configApi.get(WIDGET_CODES.taskList);
      console.log(config);

      const taskList = lazyLoading
        ? await taskListApi.get(config.connection, currentPage, 10)
        : await taskListApi.get(config.connection);
      console.log(taskList);

      this.setState({
        loading: false,
        columns: normalizeColumns(config.columns, taskList.payload[0]),
        rows: normalizeRows(taskList.payload),
        size: taskList.metadata.size,
        connection: config.connection,
      });
    } catch (error) {
      this.handleError(error);
    }
  };

  componentDidUpdate = prevProps => {
    const { lazyLoading } = this.props;
    if (prevProps.lazyLoading !== lazyLoading) {
      this.updateRows(lazyLoading ? 0 : undefined);
    }
  };

  updateRows = async (
    page,
    rowsPerPage = 10,
    sortedColumn,
    sortedOrder,
    filter,
    callback = () => {},
    withDelay
  ) => {
    const { connection } = this.state;

    if (withDelay) {
      clearTimeout(this.timer.ref);
      await utils.timeout(800, this.timer);
    }

    this.setState({ loading: true });
    try {
      const res = await taskListApi.get(
        connection,
        page,
        rowsPerPage,
        sortedColumn,
        sortedOrder,
        filter
      );

      this.setState({
        rows: normalizeRows(res.payload),
        size: res.metadata.size,
        currentPage: page || 0,
        loading: false,
      });
      callback();
    } catch (error) {
      this.handleError(error);
      this.setState({ loading: false });
    }
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleError(err) {
    const { onError } = this.props;
    onError(err);
    this.setState({
      error: true,
      errorAlert: err,
    });
  }

  render() {
    const { loading, columns, rows, currentPage, size, error, errorAlert } = this.state;
    const { classes, lazyLoading } = this.props;

    let lazyLoadingProps;
    if (lazyLoading) {
      lazyLoadingProps = {
        currentPage,
        onChange: this.updateRows,
        size,
      };
    }

    return (
      <ThemeProvider theme={theme}>
        <Paper className={classes.paper}>
          {error ? (
            <ErrorComponent message={i18next.t('messages.errors.dataLoading')} />
          ) : (
            <Table
              loading={loading}
              title={i18next.t('table.title')}
              subtitle={i18next.t('table.subtitle')}
              columns={columns}
              rows={rows}
              rowsPerPageOptions={[10, 25, 50, 100]}
              lazyLoadingProps={lazyLoadingProps}
            />
          )}
        </Paper>
        <ErrorNotification message={errorAlert} onClose={this.closeNotification} />
      </ThemeProvider>
    );
  }
}

TaskList.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
  }),
  lazyLoading: PropTypes.bool,
  onError: PropTypes.func,
};

TaskList.defaultProps = {
  classes: {},
  lazyLoading: false,
  onError: () => {},
};

export default withStyles(styles)(TaskList);
