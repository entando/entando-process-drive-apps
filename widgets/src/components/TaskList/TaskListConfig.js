import React from 'react';
import { FormGroup, ControlLabel, Checkbox, HelpBlock, Row, Col } from 'patternfly-react';
import { SortableContainer, SortableElement } from 'react-sortable-hoc';
import arrayMove from 'array-move';
import { withStyles } from '@material-ui/core/styles';

import { getConnections } from 'api/pda/connections';
import { getGroups } from 'api/pda/groups';
import { getColumns } from 'api/pda/tasks';
import { normalizeConfigColumns, normalizeConfigGroups } from 'components/TaskList/normalizeData';
import taskListConfigType from 'types/taskListConfigType';
import RenderSwitch from 'components/common/RenderSwitch';
import Notification from 'components/common/Notification';

const generalOptions = [
  {
    key: 'newPageOnClick',
    label: 'Open new page on table row click',
    checked: false,
  },
  {
    key: 'showClaim',
    label: 'Show Claim Button',
    checked: true,
  },
  {
    key: 'showComplete',
    label: 'Show Complete Button',
    checked: true,
  },
];

const SortableRow = SortableElement(({ value, onColumnChange }) => (
  <tr>
    <td>
      <i className="fa fa-bars" style={{ cursor: 'ns-resize' }} /> {value.position}
    </td>
    <td>{value.name}</td>
    <td>
      <Checkbox
        style={{ margin: 0 }}
        bsClass="checkbox"
        checked={value.isVisible}
        onChange={onColumnChange}
      />
    </td>
    <td>
      <input
        type="text"
        className="form-control"
        onChange={onColumnChange}
        value={value.header || ''}
      />
    </td>
  </tr>
));

const SortableBody = SortableContainer(({ items, onColumnChange }) => (
  <tbody>
    {items.map((value, index) => (
      <SortableRow
        key={value.name}
        index={index}
        value={value}
        onColumnChange={onColumnChange(index)}
      />
    ))}
  </tbody>
));

const styles = {
  paper: {
    padding: 20,
  },
  title: {
    display: 'flex',
    alignItems: 'center',
    paddingBottom: 15,
  },
  subtitle: {
    padding: '20px 0 20px 0',
  },
  formControl: {
    minWidth: 200,
  },
};

class TaskListConfig extends React.Component {
  state = {
    config: {
      knowledgeSource: '',
      options: generalOptions,
      groups: [],
      columns: [],
    },
    sourceList: [],
    errorAlert: null,
  };

  componentDidMount = async () => {
    try {
      // get list of connections
      const sourceList = await getConnections();
      if (sourceList.errors && sourceList.errors.length) {
        throw sourceList.errors[0];
      }

      // fetch screen if a default config prop is passed
      this.setState({ sourceList: sourceList.payload }, this.fetchScreen);
    } catch (error) {
      this.handleError(error.message);
    }
  };

  componentDidUpdate = async prevProps => {
    const { config } = this.props;

    // refetch state if config changes
    if (JSON.stringify(config) !== JSON.stringify(prevProps.config)) {
      this.fetchScreen();
    }
  };

  fetchScreen = () => {
    const { config } = this.props;
    if (config && config.knowledgeSource) {
      this.onChangeSource(config.knowledgeSource, () => {
        this.setState({
          config: {
            ...config,
            groups: JSON.parse(config.groups),
            options: JSON.parse(config.options),
            columns: JSON.parse(config.columns),
          },
        });
      });
    }
  };

  onChangeSource = (e, cb) => {
    const { config } = this.state;
    const knowledgeSource = e.target ? e.target.value : e;
    this.setState(
      {
        config: {
          ...config,
          options: [...generalOptions],
          groups: [],
          columns: [],
          knowledgeSource,
        },
      },
      async () => {
        if (cb) {
          cb();
        } else {
          try {
            const { payload: groups } = await getGroups(knowledgeSource);
            const { payload: columns } = await getColumns(knowledgeSource);

            this.setState(state => ({
              config: {
                ...state.config,
                groups: normalizeConfigGroups(groups),
                columns: normalizeConfigColumns(columns),
              },
            }));
          } catch (error) {
            // eslint-disable-next-line no-console
            console.log('Error while trying to fetch groups and columns from PDA server', error);
          }
        }
      }
    );
  };

  handleSortStart = ({ node }) => {
    const tds = document.getElementsByClassName('table-sortableHelper')[0].childNodes;
    node.childNodes.forEach((n, idx) => {
      tds[idx].style.width = `${n.offsetWidth}px`;
    });
  };

  handleSortEnd = ({ oldIndex, newIndex }) => {
    this.setState(({ config }) => ({
      config: {
        ...config,
        columns: arrayMove(config.columns, oldIndex, newIndex).map((column, i) => ({
          ...column,
          position: i,
        })),
      },
    }));
  };

  handleColumnChange = index => e => {
    const { config } = this.state;
    const { value, checked, type } = e.target;
    const newColumns = [...config.columns];
    if (type === 'checkbox') {
      newColumns[index].isVisible = checked;
    } else {
      newColumns[index].header = value;
    }
    this.setState({ config: { ...config, columns: newColumns } });
  };

  handleOptions = index => e => {
    const { config } = this.state;
    const { value } = e.state;
    const newOptions = [...config.options];

    newOptions[index].checked = value;
    this.setState({ config: { ...config, options: newOptions } });
  };

  handleGroups = index => e => {
    const { config } = this.state;
    const { value } = e.state;
    const newGroups = [...config.groups];

    newGroups[index].checked = value;
    this.setState({ config: { ...config, groups: newGroups } });
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleError(err) {
    this.setState({
      errorAlert: err,
    });
  }

  render() {
    const { sourceList, config, errorAlert } = this.state;
    const { knowledgeSource, groups, columns, options } = config;
    return (
      <div>
        <Notification type="error" message={errorAlert} onClose={this.closeNotification} />
        <form>
          <Row>
            <Col xs={12}>
              <FormGroup controlId="connection">
                <ControlLabel>Knowledge Source</ControlLabel>
                <select
                  className="form-control"
                  value={knowledgeSource}
                  onChange={this.onChangeSource}
                >
                  <option value="">Select...</option>
                  {sourceList.map(source => (
                    <option key={source.name} value={source.name}>
                      {source.name}
                    </option>
                  ))}
                </select>
                <HelpBlock>Select one of the Kie server connections.</HelpBlock>
              </FormGroup>
            </Col>
          </Row>
          {knowledgeSource && (
            <section>
              <legend>General options</legend>
              <Row>
                <Col xs={12}>
                  {options.map((item, i) => (
                    <RenderSwitch
                      key={item.key}
                      id={item.key}
                      label={item.label}
                      checked={item.checked}
                      onChange={this.handleOptions(i)}
                    />
                  ))}
                </Col>
              </Row>
              <legend>BPM Groups</legend>
              <Row>
                <Col xs={12}>
                  {groups.map((item, i) => (
                    <RenderSwitch
                      key={item.key}
                      id={item.key}
                      label={item.label}
                      checked={item.checked}
                      onChange={this.handleGroups(i)}
                    />
                  ))}
                </Col>
              </Row>
              <legend>Task Columns</legend>
              <Row>
                <Col xs={12}>
                  <table className="table">
                    <thead>
                      <tr>
                        <th>Position</th>
                        <th>Column Name</th>
                        <th>Visible</th>
                        <th>Override Fields</th>
                      </tr>
                    </thead>
                    <SortableBody
                      items={columns}
                      helperClass="table-sortableHelper"
                      onSortStart={this.handleSortStart}
                      onSortEnd={this.handleSortEnd}
                      onColumnChange={this.handleColumnChange}
                    />
                  </table>
                </Col>
              </Row>
            </section>
          )}
        </form>
      </div>
    );
  }
}

TaskListConfig.propTypes = {
  config: taskListConfigType,
};

TaskListConfig.defaultProps = {
  config: {},
};

const TaskListConfigContainer = withStyles(styles)(TaskListConfig);

export default TaskListConfigContainer;
