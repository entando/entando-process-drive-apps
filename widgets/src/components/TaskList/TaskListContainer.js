import { MuiThemeProvider as ThemeProvider } from '@material-ui/core/styles';
import React from 'react';
import i18next from 'i18next';
import PropTypes from 'prop-types';
import Paper from '@material-ui/core/Paper';
import withStyles from '@material-ui/core/styles/withStyles';
import SVG from 'react-inlinesvg';
import Toolbar from '@material-ui/core/Toolbar';
import Typography from '@material-ui/core/Typography';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

import { DOMAINS, LOCAL } from 'api/constants';
import { getTasks } from 'api/pda/tasks';
import { getDiagram } from 'api/pda/processes';
import { getPageWidget } from 'api/app-builder/pages';
import utils from 'utils';

import { normalizeColumns, normalizeRows } from 'components/TaskList/normalizeData';
import SearchInput from 'components/common/SearchInput';
import Notification from 'components/common/Notification';
import ErrorComponent from 'components/common/ErrorComponent';
import Table from 'components/common/Table/Table';
import SimpleDialog from 'components/common/SimpleDialog';
import theme from 'theme';

const styles = {
  toolbar: {
    justifyContent: 'space-between',
    padding: '16px 16px 8px 16px',
    minHeight: 'unset',
  },
  noSubtitleToolbar: {
    padding: '8px 16px',
  },
  title: {
    textAlign: 'left',
  },
  paper: {
    minHeight: 459,
    position: 'relative',
  },
};

class TaskList extends React.Component {
  state = {
    loading: true,
    columns: [],
    rows: [],
    size: 0,
    connection: {},
    filter: '',
    blocker: '',
    errorAlert: null,
    lastPage: false,
    activeTab: 0,
    diagramModal: {
      open: false,
      title: '',
      body: '',
    },
    groups: [],
    page: 0,
  };

  timer = { ref: null };

  componentDidMount = async () => {
    const { lazyLoading, serviceUrl, pageCode, frameId, onSelectTask } = this.props;

    if (!LOCAL) {
      // set the PDA domain to the URL passed via props
      DOMAINS.PDA = serviceUrl;
    }

    try {
      // config will be fetched from app-builder
      const widgetConfigs = await getPageWidget(pageCode, frameId, 'TASK_LIST');
      if (widgetConfigs.errors && widgetConfigs.errors.length) {
        throw widgetConfigs.errors[0];
      }
      if (!widgetConfigs.payload) {
        throw new Error('messages.errors.widgetConfig');
      }

      const { config } = widgetConfigs.payload;
      const connection = config.knowledgeSource;
      const groups = JSON.parse(config.groups)
        .filter(group => group.checked)
        .map(group => group.key);

      const taskList = lazyLoading
        ? await getTasks({ connection, groups }, 0, 10)
        : await getTasks({ connection, groups });

      if (!taskList.payload) {
        throw new Error('messages.errors.errorResponse');
      }

      if (!taskList.payload.length) {
        this.setState({ blocker: 'taskList.emptyList' });
      } else {
        const options = JSON.parse(config.options);
        const rows = normalizeRows(taskList.payload);

        // dispatch onSelectTask event for the first item on list
        onSelectTask({ ...rows[0], pos: 0, groups });

        this.setState({
          loading: false,
          columns: normalizeColumns(JSON.parse(config.columns), rows[0], options, {
            openDiagram: this.openDiagram,
            selectTask: onSelectTask,
          }),
          rows,
          lastPage: taskList.metadata.lastPage === 1,
          size: taskList.metadata.size,
          connection: config.knowledgeSource,
          groups,
        });
      }
    } catch (error) {
      this.handleError(error.message, 'messages.errors.dataLoading');
    }
  };

  componentDidUpdate = prevProps => {
    const { lazyLoading } = this.props;
    if (prevProps.lazyLoading !== lazyLoading) {
      this.updateRows(lazyLoading ? 0 : undefined);
    }
  };

  updateRows = async (
    page,
    rowsPerPage = 10,
    sortedColumn,
    sortedOrder,
    filter,
    callback = () => {},
    withDelay
  ) => {
    const { connection, groups } = this.state;

    if (withDelay) {
      clearTimeout(this.timer.ref);
      await utils.timeout(800, this.timer);
    }

    this.setState({ loading: true });
    try {
      const res = await getTasks(
        { connection, groups },
        page,
        rowsPerPage,
        sortedColumn,
        sortedOrder,
        filter
      );
      if (!res.payload) {
        throw res.message;
      }

      this.setState({
        rows: normalizeRows(res.payload),
        size: res.metadata.size,
        lastPage: res.metadata.lastPage === 1,
        loading: false,
      });
      callback();
    } catch (error) {
      this.handleError(error, 'messages.errors.dataLoading');
      this.setState({ loading: false });
    }
  };

  openDiagram = row => async () => {
    const { connection } = this.state;

    if (row.processId && row.id) {
      const process = `${row.processId}@${row.id.split('@')[1]}`;
      this.setState({ loading: true });
      try {
        const diagram = await getDiagram(connection, process);
        this.setState({
          diagramModal: {
            title: `${i18next.t('taskList.diagramModalTitle')}: ${process}`,
            body: <SVG src={diagram} />,
            open: true,
          },
        });
      } catch (error) {
        this.handleError(error);
      } finally {
        this.setState({ loading: false });
      }
    }
  };

  onCloseDiagramModal = () => {
    const { diagramModal } = this.state;
    this.setState({ diagramModal: { ...diagramModal, open: false } });
  };

  closeNotification = () => {
    this.setState({ errorAlert: null });
  };

  handleChangeFilter = event => {
    const { lazyLoading } = this.props;
    const { rowsPerPage, sortedColumn, sortOrder } = this.state;
    const filter = event.target.value;

    this.setState({ filter });
    if (lazyLoading) {
      this.updateRows(0, rowsPerPage, sortedColumn, sortOrder, filter, undefined, true);
    }
  };

  handleChangeTab = (_, activeTab) => {
    this.setState({ activeTab });
  };

  handleChangePage = page => {
    this.setState({
      page,
    });
  };

  handleRowSelect = (row, idx) => {
    const { onSelectTask } = this.props;
    const { page, rowsPerPage } = this.state;
    const pos = page * (rowsPerPage || 10) + idx;
    onSelectTask({ ...row, pos });
  };

  handleError(err, blocker = '') {
    const { onError } = this.props;
    onError(err);
    this.setState({
      errorAlert: err.toString(),
      blocker,
    });
  }

  render() {
    const {
      loading,
      columns,
      rows,
      size,
      blocker,
      errorAlert,
      lastPage,
      diagramModal,
      filter,
      activeTab,
    } = this.state;
    const { classes, lazyLoading } = this.props;

    let lazyLoadingProps;
    if (lazyLoading) {
      lazyLoadingProps = {
        onChange: this.updateRows,
        size,
        lastPage,
      };
    }

    return (
      <ThemeProvider theme={theme}>
        <Paper className={classes.paper}>
          {blocker ? (
            <ErrorComponent message={blocker} />
          ) : (
            <>
              <Toolbar className={classes.toolbar}>
                <div className={classes.title}>
                  <Typography variant="h2">{i18next.t('table.title')}</Typography>
                </div>
                <div>
                  <SearchInput value={filter} onChange={this.handleChangeFilter} />
                </div>
              </Toolbar>
              <Tabs
                indicatorColor="primary"
                textColor="primary"
                onChange={this.handleChangeTab}
                value={activeTab}
              >
                <Tab label={i18next.t('taskList.tabs.myTasks')} />
              </Tabs>
              <Table
                loading={loading}
                columns={columns}
                rows={rows}
                rowsPerPageOptions={[10, 25, 50, 100]}
                lazyLoadingProps={lazyLoadingProps}
                onRowClick={this.handleRowSelect}
                onChangePage={this.handleChangePage}
              />
            </>
          )}
        </Paper>
        <SimpleDialog
          open={diagramModal.open}
          title={diagramModal.title}
          body={diagramModal.body}
          onClose={this.onCloseDiagramModal}
          maxWidth="xl"
          fullWidth
        />
        <Notification type="error" message={errorAlert} onClose={this.closeNotification} />
      </ThemeProvider>
    );
  }
}

TaskList.propTypes = {
  classes: PropTypes.shape({
    paper: PropTypes.string,
    toolbar: PropTypes.string,
    title: PropTypes.string,
  }),
  lazyLoading: PropTypes.bool,
  onError: PropTypes.func,
  onSelectTask: PropTypes.func,
  serviceUrl: PropTypes.string,
  pageCode: PropTypes.string,
  frameId: PropTypes.string,
};

TaskList.defaultProps = {
  classes: {},
  lazyLoading: true,
  onError: () => {},
  onSelectTask: () => {},
  serviceUrl: '',
  pageCode: '',
  frameId: '',
};

export default withStyles(styles)(TaskList);
