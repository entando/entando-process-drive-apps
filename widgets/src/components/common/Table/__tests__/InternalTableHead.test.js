import React from 'react';
import { render } from '@testing-library/react';

import InternalTableHead from 'components/common/Table/InternalTableHead';
import { WIDGET_CONFIGS } from 'mocks/taskList/configs';
import { normalizeColumns } from 'components/TaskList/normalizeData';
import jsonRows from 'mocks/taskList/tasks';

const columns = normalizeColumns(
  JSON.parse(WIDGET_CONFIGS.payload.config.columns),
  jsonRows.payload[0]
);

describe('<InternalTableHead />', () => {
  it('renders snapshot correctly', () => {
    const columnNumber = 2;
    const { container } = render(
      <table>
        <InternalTableHead columns={columns.slice(0, columnNumber)} />
      </table>
    );

    expect(container).toMatchSnapshot();
    expect(document.querySelectorAll('th').length).toBe(columnNumber);
  });
});
