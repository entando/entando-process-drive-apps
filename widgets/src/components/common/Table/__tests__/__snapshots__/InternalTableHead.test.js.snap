// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InternalTableHead /> renders snapshot correctly 1`] = `
<div>
  <table>
    <thead
      class="MuiTableHead-root"
    >
      <tr
        class="MuiTableRow-root MuiTableRow-head"
      >
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
          style="white-space: nowrap; position: sticky; left: 0px; background: white; width: 20px; z-index: 100; border-right: 1px solid #eee; padding-right: 16px; text-align: center;"
        >
          Action
        </th>
        <th
          class="MuiTableCell-root MuiTableCell-head"
          scope="col"
          style="white-space: nowrap;"
        >
          <span
            class="MuiButtonBase-root MuiTableSortLabel-root TableSortLabel-notActive-21"
            role="button"
            tabindex="0"
          >
            id
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiTableSortLabel-icon MuiTableSortLabel-iconDirectionAsc"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M20 12l-1.41-1.41L13 16.17V4h-2v12.17l-5.58-5.59L4 12l8 8 8-8z"
              />
            </svg>
          </span>
        </th>
      </tr>
    </thead>
  </table>
</div>
`;
